File: apps/demo-be-java-vertx/src/test/java/com/demobejavx/unit/PasswordServiceTest.java
package com.demobejavx.unit;

import com.demobejavx.auth.PasswordService;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

class PasswordServiceTest {

    private PasswordService passwordService;

    @BeforeEach
    void setUp() {
        passwordService = new PasswordService();
    }

    @Test
    void hash_returnsNonNullHash() {
        String hash = passwordService.hash("Str0ng#Pass1");
        assertNotNull(hash);
    }

    @Test
    void hash_returnsHashDifferentFromPlaintext() {
        String plaintext = "Str0ng#Pass1";
        String hash = passwordService.hash(plaintext);
        assertNotEquals(plaintext, hash);
    }

    @Test
    void verify_correctPassword_returnsTrue() {
        String hash = passwordService.hash("Str0ng#Pass1");
        assertTrue(passwordService.verify("Str0ng#Pass1", hash));
    }

    @Test
    void verify_wrongPassword_returnsFalse() {
        String hash = passwordService.hash("Str0ng#Pass1");
        assertFalse(passwordService.verify("WrongPass#1", hash));
    }

    @Test
    void hash_samePasswordTwice_returnsDifferentHashes() {
        String hash1 = passwordService.hash("Str0ng#Pass1");
        String hash2 = passwordService.hash("Str0ng#Pass1");
        // BCrypt salts are random, so different hashes are expected
        assertNotEquals(hash1, hash2);
    }
}

File: apps/demo-be-ktkt/src/test/kotlin/com/organiclever/demoktkt/unit/PasswordServiceTest.kt
package com.organiclever.demoktkt.unit

import com.organiclever.demoktkt.auth.PasswordService
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertNotEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

class PasswordServiceTest {
  private val service = PasswordService()

  @Test
  fun `hash produces non-empty string`() {
    val hash = service.hash("Str0ng#Pass1")
    assertTrue(hash.isNotEmpty())
  }

  @Test
  fun `hash is different from plaintext`() {
    val password = "Str0ng#Pass1"
    val hash = service.hash(password)
    assertNotEquals(password, hash)
  }

  @Test
  fun `verify returns true for correct password`() {
    val password = "Str0ng#Pass1"
    val hash = service.hash(password)
    assertTrue(service.verify(password, hash))
  }

  @Test
  fun `verify returns false for wrong password`() {
    val hash = service.hash("Str0ng#Pass1")
    assertFalse(service.verify("WrongPass#1234", hash))
  }

  @Test
  fun `two hashes of same password are different (salt)`() {
    val password = "Str0ng#Pass1"
    val hash1 = service.hash(password)
    val hash2 = service.hash(password)
    assertNotEquals(hash1, hash2)
  }

  @Test
  fun `verify returns false for malformed hash`() {
    // BCrypt.checkpw throws an exception for malformed hashes; getOrDefault covers this
    assertFalse(service.verify("anyPassword", "not-a-valid-bcrypt-hash"))
  }
}
